src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs

```diff
@@ -193,14 +208,25 @@ private IISAwaitable DoFlushAsync()
             unsafe
             {
                 var hr = 0;
-                hr = NativeMethods.http_flush_response_bytes(_pHttpContext, IISAwaitable.FlushCallback, (IntPtr)_thisHandle, out var fCompletionExpected);
-
-                if (!fCompletionExpected)
+                if (_wasUpgraded)
                 {
-                    CompleteFlush(hr, 0);
+                    hr = NativeMethods.http_websockets_flush_bytes(_pHttpContext, IISAwaitable.FlushCallback, (IntPtr)_thisHandle, out var fCompletionExpected);
+                    if (!fCompletionExpected)
+                    {
+                        CompleteFlushWebSockets(hr, 0);
+                    }
+                    return _flushWebSocketsOperation;
+                }
+                else
+                {
+                    hr = NativeMethods.http_flush_response_bytes(_pHttpContext, out var fCompletionExpected);
+                    if (!fCompletionExpected)
+                    {
+                        OnAsyncCompletion(hr, 0);
```

Comment:
So this is code reuse but you know what the operation is so you can just call complete on the `_operation directly.